### PR TITLE
[Feature][Flink] Support flow control in Flink

### PR DESCRIPTION
**Pull request by zhilinli123**
## Check list
SubTask:https://github.com/apache/seatunnel/issues/5503

* [ ] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/seatunnel/blob/dev/seatunnel-dist/pom.xml)
* [ ] Update the [`release-note`](https://github.com/apache/seatunnel/blob/dev/release-note.md).

### CODE REVIEW

**Comment by zhilinli123** [comment]
PTAL : @Hisoka-X @hailin0

**Comment by TyrantLucifer** [comment]
Please offer the test cases for your change to verify it worked.

**Comment by zhilinli123** [comment]
Pasted Graphic

**Comment by zhilinli123** [comment]
image

**Comment by zhilinli123** [comment]
PTAL: @TyrantLucifer @hailin0 @Hisoka-X  thks

File: seatunnel-core/seatunnel-flink-starter/seatunnel-flink-starter-common/src/main/java/org/apache/seatunnel/core/starter/flink/execution/SourceExecuteProcessor.java [hunk 1]
```diff
@@ -52,10 +53,11 @@
 
 public class SourceExecuteProcessor extends FlinkAbstractPluginExecuteProcessor<SeaTunnelSource> {
     private static final String PLUGIN_TYPE = PluginType.SOURCE.getType();
+    private Config envConfis;
 
-    public SourceExecuteProcessor(
-            List<URL> jarPaths, List<? extends Config> sourceConfigs, JobContext jobContext) {
-        super(jarPaths, sourceConfigs, jobContext);
+    public SourceExecuteProcessor(List<URL> jarPaths, Config ConfigsInfo, JobContext jobContext) {
```
Comment (Hisoka-X) [inline seatunnel-core/seatunnel-flink-starter/seatunnel-flink-starter-common/src/main/java/org/apache/seatunnel/core/starter/flink/execution/SourceExecuteProcessor.java:59]:
```suggestion
    public SourceExecuteProcessor(List<URL> jarPaths, Config configsInfo, JobContext jobContext) {
```

File: seatunnel-core/seatunnel-flink-starter/seatunnel-flink-starter-common/src/main/java/org/apache/seatunnel/core/starter/flink/execution/SourceExecuteProcessor.java [hunk 2]
```diff
@@ -68,10 +70,11 @@ public List<DataStream<Row>> execute(List<DataStream<Row>> upstreamDataStreams)
             Config pluginConfig = pluginConfigs.get(i);
             BaseSeaTunnelSourceFunction sourceFunction;
             if (internalSource instanceof SupportCoordinate) {
-                sourceFunction = new SeaTunnelCoordinatedSource(internalSource);
+                sourceFunction = new SeaTunnelCoordinatedSource(internalSource, envConfis);
```
Comment (Hisoka-X) [inline seatunnel-core/seatunnel-flink-starter/seatunnel-flink-starter-common/src/main/java/org/apache/seatunnel/core/starter/flink/execution/SourceExecuteProcessor.java]:
```suggestion
                sourceFunction = new SeaTunnelCoordinatedSource(internalSource, envConfigs);
```

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/task/SourceSeaTunnelTask.java [hunk 3]
```diff
@@ -47,7 +47,7 @@ public class SourceSeaTunnelTask<T, SplitT extends SourceSplit> extends SeaTunne
 
     private transient Object checkpointLock;
     @Getter private transient Serializer<SplitT> splitSerializer;
-    private final Map<String, Object> envOption;
+    private final transient Map<String, Object> envOption;
```
Comment (Hisoka-X) [inline seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/task/SourceSeaTunnelTask.java]: Any reason for add `transient`?

File: seatunnel-translation/seatunnel-translation-flink/seatunnel-translation-flink-common/src/main/java/org/apache/seatunnel/translation/flink/source/BaseSeaTunnelSourceFunction.java [hunk 4]
```diff
@@ -63,12 +68,25 @@ public abstract class BaseSeaTunnelSourceFunction extends RichSourceFunction<Row
 
     protected final AtomicLong latestCompletedCheckpointId = new AtomicLong(0);
     protected final AtomicLong latestTriggerCheckpointId = new AtomicLong(0);
+    // Env Conf Info
+    private Config envConfis;
```
Comment (Hisoka-X) [inline seatunnel-translation/seatunnel-translation-flink/seatunnel-translation-flink-common/src/main/java/org/apache/seatunnel/translation/flink/source/BaseSeaTunnelSourceFunction.java]: ditto

File: seatunnel-translation/seatunnel-translation-flink/seatunnel-translation-flink-common/src/main/java/org/apache/seatunnel/translation/flink/source/BaseSeaTunnelSourceFunction.java [hunk 5]
```diff
@@ -63,12 +68,25 @@ public abstract class BaseSeaTunnelSourceFunction extends RichSourceFunction<Row
...
+    // Store env for external Settings
+    private Map<String, Object> envOption = new HashMap<>();
 
     /** Flag indicating whether the consumer is still running. */
     private volatile boolean running = true;
 
-    public BaseSeaTunnelSourceFunction(SeaTunnelSource<SeaTunnelRow, ?, ?> source) {
+    public BaseSeaTunnelSourceFunction(
+            SeaTunnelSource<SeaTunnelRow, ?, ?> source, Config sourceEnvConfis) {
```
Comment (Hisoka-X) [inline seatunnel-translation/seatunnel-translation-flink/seatunnel-translation-flink-common/src/main/java/org/apache/seatunnel/translation/flink/source/BaseSeaTunnelSourceFunction.java]: ditto

File: seatunnel-translation/seatunnel-translation-flink/seatunnel-translation-flink-common/src/main/java/org/apache/seatunnel/translation/flink/source/SeaTunnelCoordinatedSource.java [hunk 6]
```diff
@@ -27,9 +29,10 @@ public class SeaTunnelCoordinatedSource extends BaseSeaTunnelSourceFunction {
 
     protected static final String COORDINATED_SOURCE_STATE_NAME = "coordinated-source-states";
 
-    public SeaTunnelCoordinatedSource(SeaTunnelSource<SeaTunnelRow, ?, ?> source) {
+    public SeaTunnelCoordinatedSource(
+            SeaTunnelSource<SeaTunnelRow, ?, ?> source, Config envConfis) {
         // TODO: Make sure the source is coordinated.
-        super(source);
+        super(source, envConfis);
```
Comment (Hisoka-X) [inline seatunnel-translation/seatunnel-translation-flink/seatunnel-translation-flink-common/src/main/java/org/apache/seatunnel/translation/flink/source/SeaTunnelCoordinatedSource.java]: ditto

File: seatunnel-translation/seatunnel-translation-flink/seatunnel-translation-flink-common/src/main/java/org/apache/seatunnel/translation/flink/source/SeaTunnelParallelSource.java [hunk 7]
```diff
@@ -31,9 +33,9 @@ public class SeaTunnelParallelSource extends BaseSeaTunnelSourceFunction
 
     protected static final String PARALLEL_SOURCE_STATE_NAME = "parallel-source-states";
 
-    public SeaTunnelParallelSource(SeaTunnelSource<SeaTunnelRow, ?, ?> source) {
+    public SeaTunnelParallelSource(SeaTunnelSource<SeaTunnelRow, ?, ?> source, Config envConfis) {
         // TODO: Make sure the source is uncoordinated.
-        super(source);
+        super(source, envConfis);
```
Comment (Hisoka-X) [inline seatunnel-translation/seatunnel-translation-flink/seatunnel-translation-flink-common/src/main/java/org/apache/seatunnel/translation/flink/source/SeaTunnelParallelSource.java]: ditto

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/task/SourceSeaTunnelTask.java (hunk 3 above)
Comment (zhilinli123) [reply seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/task/SourceSeaTunnelTask.java]: Sorry for my oversight. Thank you for your review. I will make some changes

File: seatunnel-translation/seatunnel-translation-flink/seatunnel-translation-flink-common/src/main/java/org/apache/seatunnel/translation/flink/source/SeaTunnelParallelSource.java (hunk 7 above)
Comment (zhilinli123) [reply seatunnel-translation/seatunnel-translation-flink/seatunnel-translation-flink-common/src/main/java/org/apache/seatunnel/translation/flink/source/SeaTunnelParallelSource.java]: done

**Review by Hisoka-X — APPROVED** [review]
Overall LGTM

**Review by EricJoy2048 — APPROVED** [review]

**Pull request merged**